Bare Metal Mantl

larry-svds opened the pull request:
This is largely a documentation pull request. It documents how to install mantl on pre existing centos 7.2 instances, such as a set of bare-metal machines with Centos 7.2 installed. 

In order for this to work there are only two places where current roles need to change. 

1. When you include lvm.yml in role docker it installs cloud-utils-grow-part.  This can not be installed on bare-metal.  
2. In role/traefik/templates/traefik.toml.j2 the networkInterface is hard coded to "eth0" which apparently works on all clouds.  On hardware this is often not true.  @lastlee came up with the idea of changing this from:

        networkInterface = eth0 

to

      networkInterface = "{{ ansible_default_ipv4['interface'] }}"

The one down side of this change is that in the main yml file for role=edge needs to set 

      gather_facts: yes

I think this is a great solution but it does affect transform.samle.yml,  the alternative is to create a variable with eth0 as the default so that only bare-metal.yml would have to specify the ethernet interface for the edge nodes.

siddharthist commented on docs/bare-metal/bare-metal.yml:
```diff
@@ -0,0 +1,106 @@
+---
```
Since this is largely the same as terraform.sample.yml, we should figure out how to make them work nicely together, rather than provide a separate file. Maybe just `-e provider=bare-metal` would be enough? Maybe with some of those other variables?

larry-svds commented:
@siddharthist That totally makes sense.  and the other variables are things that are choices anyways.  **We still need to address the fact that the traefik role has the network interface hard coded to eth0** but the rest of what is in bare-metal.yml could be handled with. 

 * -e provider=bare-metal
 * -e docker-lvm-backed=true 
   * The bare-metal user has to prepare partitions for this and might not want to, but I provided documentation on how to do it and it is definitely recomended on centos. 
 * -e docker_lvm_data_volume_size="80%FREE"
   * This just points out that if you are not sharing this volume between docker and gluster there isn't much point in setting it to 40%FREE.  I'd love some feed back on if my understanding is correct on that. 

And then there is how mantl should handle edge.  @lastlee's solution using gathered facts would work even in cases where the edge machines have different networkInterfaces. For it to work though, `terraform.sample.yml` has to change, even for the pull request as submitted.  This is because this pull also changes the template `roles/traefik/templates/traefik.toml.j2` under the implicit assumption that the facts are being gathered.  

An alternative is to input the networkInterface as an environment varaible and then have a default in `roles/traefik/default/main.yml of eth0. If that is the route we should go, then I'd modify this pull to use a variable and document: 

 * -e traefik_network_interface=ensp0
   * based on the interface card in the edge node hardware.

siddharthist commented:
@larry-svds I like the idea of gathering facts on edge nodes and just documenting the installation with command-line options. That Traefik config should make our whole setup more flexible and reliable. Should be much simpler :) thanks for your hard work on bare-metal support!

larry-svds commented:
got rid of bare-metal.yml, updated the docs to simply run against terraform.sample.yml with environment variables, AND changed gather facts in terraform.sample.yml.

siddharthist commented on roles/docker/tasks/main.yml:
```diff
@@ -109,6 +109,7 @@
 
 - include: lvm.yml
   when: docker_lvm_backed|bool
+  and: provider!=bare-metal
```
I'm not sure this is valid Ansible syntax. You might want:
```yaml
    when: docker_lvm_backend|bool and provider != bare-metal
```
It also looks like the conditional has a typo right now (`backed->backend`), so we might want to fix that too!

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+By bare-metal we mean a set of physical computers that
+have centos 7 installed on them.
+
+If you are using open-stack, Vmware, or a cloud provider, Mantl.io has terraform scripts for
```
Openstack, VMWare

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+instances have 1 core and 4GB RAM).  In other words, your machines could
+be even smaller than our test system.   In fact, maybe you
+have a couple of physical boxes, and then created a few virtual machine vms for the controls
+on a couple of laptops. They have also been used on much larger machines.
```
I'm not sure much information on hardware is necessary here, other than to state some reasonable minimum requirements (like what `m3.medium` specifies).

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+   * Remove existing partitions
+   * Press the button to automatically partition. This will give you a default set.
+   * It will put 50 toward root and the rest in home, change this:
+     * These are services machines and shouldn't really have much if in home.
```
Maybe "These machines won't have much data in /home" or "These machines won't store many files in /home"

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+   * It will put 50 toward root and the rest in home, change this:
+     * These are services machines and shouldn't really have much if in home.
+     * You will need to leave unformatted space on the drive for docker. Mantl defaults to 100 Gigs.
+     * if you have lots of space leave 50 in home, but you won't need much there. Leave the rest unformatted.
```
This seems like a duplicate of the above bullet points

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+     * You will need to leave unformatted space on the drive for docker. Mantl defaults to 100 Gigs.
+     * if you have lots of space leave 50 in home, but you won't need much there. Leave the rest unformatted.
+ * turn on your wired internet connection.  It should just be a toggle switch for your device.
+ * Once the istall starts it asks for a root password and a first user.
```
install

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+Once rebooted,
+if you forgot to turn on your internet in the install, you can set it up like so
+http://www.krizna.com/centos/setup-network-centos-7/ .  It might be easier and more
+automated (there for less error prone) to just reinstall and remember to turn on your internet during the install.
```
therefore

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+
+You could also put the dns lines in /etc/sysconfig/network.
+
+permanently change your host name with
```
hostname

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+ * Command: n
+ * partion : default
+   * please note which partition it is in.  So if its partition 5, eventually you will need to tell mantl /dev/sda5 for the LVM
+   * you kinda want all your machines to use the same partition because this partition is entered as a ssytem wide variable.
```
system

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+
+Note that I am creating a partion size 50 Gigs, this is for docker.  Just make it consistent across your cluster.
+
+if you have MS-dos partions make it a primary partion, not an extended partion or it won't work.
```
Make which partition primary? The MS-DOS ones? Or the ones we made just above?

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+
+## Creating Your Inventory
+
+Here is an example inventory file. It should be place in the root of the mantl directory.
```
placed

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+    edge01
+    edge02
+
+I had to add the ansible_ssh_host line to run `playbooks/reboot-hosts.yml`
```
Should we make this part of the PR? If it is not suitable for non-bare-metal use cases, we could just add a `when: if provider is defined and provider == "bare-metal"`.

larry-svds replied:
Let me clarify that I had to add `ansible_ssh_host` elements to this inventory (which is only used with bare-metal) in oder to be able to run the reboot-hosts.yml against this bare metal inventory. No changes were made to any thing that would be used by none bare-metal use-cases.

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+
+    ansible all -i inventory  -u centos -k -m authorized_key -a "user=centos key=https://github.com/youraccount.keys"
+
+Note this makes use of your public key on github.  if you don't have a git hub account or a key pair on your git hub
```
Github account

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+    ansible all -i inventory  -u centos -k -m authorized_key -a "user=centos key=https://github.com/youraccount.keys"
+
+Note this makes use of your public key on github.  if you don't have a git hub account or a key pair on your git hub
+account, get one.
```
Is there not just a way to collect the public key from the user's provisioning machine?

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+Note this makes use of your public key on github.  if you don't have a git hub account or a key pair on your git hub
+account, get one.
+
+the -k is needed cause I have to have it ask for a password at this point.
```
It'd be best to avoid the first person here. Additionally, this is a little unclear: are you asking Ansible for a password, or is Ansible asking you for a password?

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+
+the -k is needed cause I have to have it ask for a password at this point.
+
+now all commands can happen with out the password and -k option. Test with:
```
without

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+        172.16.222.6 control01
+        172.16.222.7 control02
+        172.16.222.8 control03
+        172.16.222.11 resource01
```
The terraform files call these nodes workers, and we might want to keep that nomenclature consistent. Terraform also adds a dash, like `worker-01`.

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+        172.16.222.16 edge01
+        172.16.222.17 edge02
+
+Copy the /etc/host file over.
```
/etc/hosts

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+    ansible all -i inventory -u centos --sudo --ask-sudo-pass -m copy -a "src=hosts dest=/etc/hosts"
+
+
+Might as well set the timezone just for grins.
```
This will be done in the main install, there's no need to do it here.

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+It asks for one admin password. At the end of that run there will be a `security.yml` file.  It will have the password
+you entered and a lot of keys and such that are used in the installs and in the installed software.
+
+The file you will be running is `terraform.sample.yml`.  Since you created your own inventory and didn't use terraform,
```
Maybe "the playbook"?

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+you entered and a lot of keys and such that are used in the installs and in the installed software.
+
+The file you will be running is `terraform.sample.yml`.  Since you created your own inventory and didn't use terraform,
+there are a few environment variables you need to set for your run.
```
These aren't environment variables, just "variables"

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+ * ansible-playbook -u centos -i inventory
+   * run the ansible play book as centos user against the inventory found in the ./inventory file.
+ * -e provider=bare-metal
+   * The "provider" is not terraform for a cloud, but rather bare-metal. If the inventory had been generated
```
The provider is never terraform -- it is one of "gce", "aws", etc. Maybe say "The provider is not one of the cloud providers, but rather bare-metal."

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+   * run the ansible play book as centos user against the inventory found in the ./inventory file.
+ * -e provider=bare-metal
+   * The "provider" is not terraform for a cloud, but rather bare-metal. If the inventory had been generated
+   by terraform on Google Cloud, this value would have been set automatically to 'gcs'
```
The inventory is generated by [terraform.py](https://github.com/CiscoCloud/terraform.py), not terraform itself.

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+ * -e consul_dc=dc1
+   * This is the name found in your ./inventory file for your datacenter.
+ * -e docker-lvm-backed=true
+   * lvm back docker is a really good idea in centos. This is why you craeted the extra partion during installation.
```
backed

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,298 @@
+   progress as ansible works through the rolls accross your inventory.
+
+
+Once you are done go to the browser and go to the /ui directory of any control node and you should see the mantlui.
```
Maybe provide an example based on the IPs above? Like "navigate to the IP of one of your control nodes, e.g. 172.16.222.6".

siddharthist commented:
@larry-svds I went ahead and did a more full review. Thanks for your work on this, we've had a ton of interest in bare-metal installations and it will be great to have documentation.

siddharthist commented:
This PR will fix #1147

larry-svds commented:
Sorry I took so long to get these edits in.  If I am not mistaken the conflict is that terraform.sample.yml has been moved?

larry-svds commented:
It looks like the file is now called sample.yml. The change I made to that file was to  gather facts for edge.

siddharthist commented on docs/bare-metal/README.md:
```diff
@@ -0,0 +1,301 @@
+By bare-metal we mean a set of physical computers that
+have centos 7 installed on them.
+
+If you are using open-stack, Vmware, or some other cloud provider, Mantl.io has terraform scripts for
```
Openstack, VMWare

lastlee commented:
For physical hosts, we may want to also consider extending the ansible timeout for reboot-hosts.yml to allow for the longer rebooting hosts to not fail the play.

larry-svds commented:
Changed the name of the file to sample.yml within the readme.  I couldn't see what lines the comments were about with these last copy edits. I reread this document and clean up quite a few things. This was basically notes as I worked through issues.  My apologies for not cleaning it up before doing the pull.  It should be clearer now.

larry-svds commented:
Did some scary git-nastics and got this PR to edit the file which is now sample.yml, rather than its previous name terraform.sample.yml. Now it has no conflicts with the base.  It should be ready to merge.

siddharthist commented:
@lastlee Could you make an issue (or even better, a PR) for that case?
